File: list/java/ArrayList.java
import java.util.Iterator;
import java.util.ListIterator;
import java.util.NoSuchElementException;
import java.util.OptionalInt;
import java.util.function.IntBinaryOperator;
import java.util.function.IntConsumer;
import java.util.function.IntPredicate;
import java.util.function.IntUnaryOperator;

public class ArrayList implements List {
    private static final int INITIAL_CAPACITY = 16;
    private static final int INSERTION_SORT_MIN = 16;

    private int[] array;
    private int elements;

    public ArrayList(int initialCapacity) {
        if (initialCapacity <= 0) {
            throw new IllegalArgumentException("Invalid initial capacity: " + String.valueOf(initialCapacity));
        }
        this.array = new int[initialCapacity];
        this.elements = 0;
    }

    public ArrayList() {
        this(INITIAL_CAPACITY);
    }

    @Override
    public void insert(int val, int index) {
        if (index > elements) {
            throw new IndexOutOfBoundsException(String.valueOf(index));
        }

        if (elements == array.length) {
            int[] newArray = new int[array.length * 2];
            System.arraycopy(array, 0, newArray, 0, array.length);
            array = newArray;
        }

        for (int i = elements; i > index; i--) {
            array[i] = array[i - 1];
        }

        array[index] = val;
        elements++;
    }

    @Override
    public void pushLeft(int val) {
        insert(val, 0);
    }

    @Override
    public void pushRight(int val) {
        insert(val, elements);
    }

    @Override
    public int remove(int index) {
        if (index >= elements) {
            throw new IndexOutOfBoundsException(String.valueOf(index));
        }

        int val = array[index];

        for (int i = index + 1; i < elements; i++) {
            array[i - 1] = array[i];
        }

        elements--;
        return val;
    }

    @Override
    public int popLeft() {
        return remove(0);
    }

    @Override
    public int popRight() {
        return remove(elements - 1);
    }

    @Override
    public void set(int val, int index) {
        if (index >= elements) {
            throw new IndexOutOfBoundsException(String.valueOf(index));
        }
        array[index] = val;
    }

    @Override
    public int get(int index) {
        if (index >= elements) {
            throw new IndexOutOfBoundsException(String.valueOf(index));
        }
        return array[index];
    }

    @Override
    public OptionalInt max() {
        if (elements == 0) {
            return OptionalInt.empty();
        }

        int val = array[0];

        for (int i = 1; i < elements; i++) {
            if (array[i] > val) {
                val = array[i];
            }
        }

        return OptionalInt.of(val);
    }

    @Override
    public OptionalInt min() {
        if (elements == 0) {
            return OptionalInt.empty();
        }

        int val = array[0];

        for (int i = 1; i < elements; i++) {
            if (array[i] < val) {
                val = array[i];
            }
        }

        return OptionalInt.of(val);
    }

    @Override
    public OptionalInt reduce(IntBinaryOperator function) {
        return reduce(0, function);
    }

    @Override
    public OptionalInt reduce(int identity, IntBinaryOperator function) {
        if (elements == 0) {
            return OptionalInt.empty();
        }

        int accumulator = identity;
        for (int i : this) {
            accumulator = function.applyAsInt(accumulator, i);
        }

        return OptionalInt.of(accumulator);
    }

    @Override
    public int findFirst(int val) {
        return findFirst(val, 0, elements - 1);
    }

    @Override
    public int findFirst(int val, int start, int end) {
        if (start > end || end >= elements) {
            throw new IllegalArgumentException();
        }

        for (; start <= end; start++) {
            if (val == array[start]) {
                return start;
            }
        }

        return -1;
    }

    @Override
    public int findLast(int val) {
        return findLast(val, 0, elements - 1);
    }

    @Override
    public int findLast(int val, int start, int end) {
        if (start > end || end >= elements) {
            throw new IllegalArgumentException();
        }

        for (; end >= start; end--) {
            if (val == array[end]) {
                return end;
            }
        }

        return -1;
    }

    @Override
    public boolean anyMatch(IntPredicate predicate) {
        for (int i : this) {
            if (predicate.test(i)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean allMatch(IntPredicate predicate) {
        for (int i : this) {
            if (!predicate.test(i)) {
                return false;
            }
        }
        return true;
    }

    @Override
    public void filter(IntPredicate predicate) {
        ListIterator<Integer> itr = listIterator();
        while (itr.hasNext()) {
            if (!predicate.test(itr.next())) {
                itr.remove();
            }
        }
    }

    @Override
    public void map(IntUnaryOperator function) {
        for (int i = 0; i < elements; i++) {
            array[i] = function.applyAsInt(array[i]);
        }
    }

    @Override
    public void forEach(IntConsumer consumer) {
        for (int i : this) {
            consumer.accept(i);
        }
    }

    private void introSort(int low, int high, int level) {
        if (low >= high) {
            return;
        }

        int length = high - low + 1;

        if (length <= INSERTION_SORT_MIN) {
            insertionSort(low, high);
        }
        else if (level == 0) {
            heapSort(low, high);
        }
        else {
            int pivot = quickSort(low, high);
            introSort(low, pivot - 1, level - 1);
            introSort(pivot + 1, high, level - 1);
        }
    }

    private void insertionSort(int low, int high) {
        for (int i = low; i <= high; i++) {
            int val = array[i];
            int j = i - 1;

            while (j >= low && array[j] > val) {
                array[j + 1] = array[j];
                j--;
            }

            array[j + 1] = val;
        }
    }

    private int quickSort(int low, int high) {
        int pivot = array[high];
        int pos = low;

        for (int i = low; i < high; i++) {
            if (array[i] < pivot) {
                Utils.swap(array, pos, i);
                pos++;
            }
        }

        Utils.swap(array, pos, high);
        return pos;
    }

    private void siftDown(int high, int i) {
        int major = i;
        int left = 2 * i + 1;
        int right = 2 * i + 2;

        if (left < high && array[left] > array[major]) {
            major = left;
        }
        if (right < high && array[right] > array[major]) {
            major = right;
        }

        if (i != major) {
            Utils.swap(array, i, major);
            siftDown(high, major);
        }
    }

    private void heapSort(int low, int high) {
        int length = high - low;

        for (int i = low + (length / 2) - 1; i >= low; i--) {
            siftDown(high, i);
        }

        int i = high;
        while (i >= low) {
            Utils.swap(array, low, i);
            siftDown(i, low);
            i--;
        }
    }

    @Override
    public void sort() {
        introSort(0, elements - 1, (int) (Math.log(elements) / Math.log(2)));
    }

    @Override
    public void reverse() {
        if (elements <= 1) {
            return;
        }

        for (int i = 0; i < elements / 2; i++) {
            int tmp = array[i];
            array[i] = array[elements - i - 1];
            array[elements - i - 1] = tmp;
        }
    }

    @Override
    public int size() {
        return elements;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (!(obj instanceof ArrayList l) || elements != l.elements) {
            return false;
        }

        ArrayList other = (ArrayList) obj;
        for (int i = 0; i < elements; i++) {
            if (array[i] != other.array[i]) {
                return false;
            }
        }

        return true;
    }

    @Override
    public String toString() {
        StringBuilder s = new StringBuilder(elements * 2);
        s.append('[');

        for (int i = 0; i < elements; i++) {
            s.append(array[i]);
            if (i < elements - 1) {
                s.append(", ");
            }
        }

        s.append(']');
        return s.toString();
    }

    @Override
    public Iterator<Integer> iterator() {
        return new Itr();
    }

    public ListIterator<Integer> listIterator() {
        return new ListItr();
    }

    private class Itr implements Iterator<Integer> {
        int cursor = -1;
        int lastReturn = -1;

        @Override
        public boolean hasNext() {
            return cursor + 1 < ArrayList.this.elements;
        }

        @Override
        public Integer next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            cursor++;
            lastReturn = cursor;
            return ArrayList.this.array[cursor];
        }

        @Override
        public void remove() {
            if (lastReturn < 0) {
                throw new IllegalStateException();
            }
            ArrayList.this.remove(lastReturn);
            lastReturn = -1;
            cursor--;
        }
    }

    private class ListItr implements ListIterator<Integer> {
        int cursor = -1;
        int lastReturn = -1;

        @Override
        public boolean hasNext() {
            return cursor + 1 < ArrayList.this.elements;
        }

        @Override
        public Integer next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            cursor++;
            lastReturn = cursor;
            return ArrayList.this.array[cursor];
        }

        @Override
        public boolean hasPrevious() {
            return cursor >= 0;
        }

        @Override
        public Integer previous() {
            if (!hasPrevious()) {
                throw new NoSuchElementException();
            }
            cursor--;
            lastReturn = cursor;
            return ArrayList.this.array[cursor];
        }

        @Override
        public int nextIndex() {
            return cursor + 1;
        }

        @Override
        public int previousIndex() {
            return cursor;
        }

        @Override
        public void remove() {
            if (lastReturn < 0) {
                throw new IllegalStateException();
            }
            ArrayList.this.remove(lastReturn);
            lastReturn = -1;
            cursor--;
        }

        @Override
        public void set(Integer e) {
            if (lastReturn < 0) {
                throw new IllegalStateException();
            }
            ArrayList.this.array[lastReturn] = e;
        }

        @Override
        public void add(Integer e) {
            ArrayList.this.insert(e, cursor + 1);
            cursor++;
        }
    }
}

File: list/c/arraylist.c
#include "list.h"
#include <stdlib.h>

#define INSERTION_SORT_MIN 16
#define FILTER_INPLACE 16

struct ArrayList {
  int *array;
  size_t allocated;
  size_t elements;
};

ArrayList* arraylist_create(size_t initial_capacity) {
  ArrayList *new = (ArrayList*) malloc(sizeof(ArrayList));
  if (new == NULL) {
    return NULL;
  }

  new->allocated = initial_capacity != 0 ? initial_capacity : 16;
  new->elements = 0;

  new->array = (int*) malloc(new->allocated * sizeof(int));
  if (new->array == NULL) {
    free(new);
    return NULL;
  }

  return new;
}

static bool increase(ArrayList *list) {
  int *new = (int*) realloc(list->array, list->allocated * 2 * sizeof(int));
  if (new == NULL) {
    return false;
  }

  list->array = new;
  list->allocated *= 2;

  return true;
}

bool arraylist_insert(ArrayList *list, size_t index, int val) {
  if (index > list->elements) {
    return false;
  }

  if (list->elements == list->allocated && !increase(list)) {
    return false;
  }

  for (size_t i = list->elements; i > index; i--) {
    list->array[i] = list->array[i - 1];
  }

  list->array[index] = val;
  list->elements++;

  return true;
}

bool arraylist_pushleft(ArrayList *list, int val) {
  return arraylist_insert(list, 0, val);
}

bool arraylist_pushright(ArrayList *list, int val) {
  return arraylist_insert(list, list->elements, val);
}

bool arraylist_remove(ArrayList *list, size_t index, int *ret) {
  if (index >= list->elements) {
    return false;
  }

  if (ret != NULL) {
    *ret = list->array[index];
  }

  for (size_t i = index + 1; i < list->elements; i++) {
    list->array[i - 1] = list->array[i];
  }

  list->elements--;

  return true;
}

bool arraylist_popleft(ArrayList *list, int *ret) {
  return arraylist_remove(list, 0, ret);
}

bool arraylist_popright(ArrayList *list, int *ret) {
  return list->elements > 0 && arraylist_remove(list, list->elements - 1, ret);
}

bool arraylist_set(ArrayList *list, size_t index, int val) {
  if (index >= list->elements) {
    return false;
  }
  list->array[index] = val;
  return true;
}

bool arraylist_get(ArrayList *list, size_t index, int *ret) {
  if (index >= list->elements) {
    return false;
  }
  *ret = list->array[index];
  return true;
}

bool arraylist_max(ArrayList *list, int *ret) {
  for (size_t i = 0; i < list->elements; i++) {
    if (i == 0 || list->array[i] > *ret) {
      *ret = list->array[i];
    }
  }
  return list->elements != 0;
}

bool arraylist_min(ArrayList *list, int *ret) {
  for (size_t i = 0; i < list->elements; i++) {
    if (i == 0 || list->array[i] < *ret) {
      *ret = list->array[i];
    }
  }
  return list->elements != 0;
}

bool arraylist_equals(ArrayList *first, ArrayList *second) {
  if (first->elements != second->elements) {
    return false;
  }

  for (size_t i = 0; i < first->elements; i++) {
    if (first->array[i] != second->array[i]) {
      return false;
    }
  }

  return true;
}

bool arraylist_reduce(ArrayList *list, int indentity, int (*function)(int val1, int val2), int *ret) {
  if (list->elements == 0) {
    return false;
  }

  *ret = indentity;
  for (size_t i = 0; i < list->elements; i++) {
    *ret = function(*ret, list->array[i]);
  }

  return true;
}

bool arraylist_anymatch(ArrayList *list, bool (*predicate)(int val)) {
  for (size_t i = 0; i < list->elements; i++) {
    if (predicate(list->array[i])) {
      return true;
    }
  }
  return false;
}

bool arraylist_allmatch(ArrayList *list, bool (*predicate)(int val)) {
  for (size_t i = 0; i < list->elements; i++) {
    if (!predicate(list->array[i])) {
      return false;
    }
  }
  return true;
}

static void filter_inplace(ArrayList *list, bool (*predicate)(int val)) {
  for (size_t i = 0; i < list->elements; i++) {
    if (!predicate(list->array[i])) {
      arraylist_remove(list, i, NULL);
    }
  }
}

static void filter_outofplace(ArrayList *list, bool (*predicate)(int val)) {
  int *new = (int*) malloc(list->allocated * sizeof(int));
  if (new == NULL) {
    filter_inplace(list, predicate);
    return;
  }

  size_t elements = 0;

  for (size_t i = 0; i < list->elements; i++) {
    if (predicate(list->array[i])) {
      new[elements] = list->array[i];
      elements++;
    }
  }

  free(list->array);

  list->array = new;
  list->elements = elements;
}

void arraylist_filter(ArrayList *list, bool (*predicate)(int val)) {
  (list->elements <= FILTER_INPLACE ? filter_inplace : filter_outofplace)(list, predicate);
}

void arraylist_map(ArrayList *list, int (*function)(int val)) {
  for (size_t i = 0; i < list->elements; i++) {
    list->array[i] = function(list->array[i]);
  }
}

void arraylist_foreach(ArrayList *list, void (*consumer)(int val)) {
  for (size_t i = 0; i < list->elements; i++) {
    consumer(list->array[i]);
  }
}

static void insertionsort(int *array, size_t length) {
  for (size_t i = 1; i < length; i++) {
    for (size_t j = i; j > 0 && array[j] < array[j - 1]; j--) {
      swap(&array[j], &array[j - 1]);
    }
  }
}

static void siftdown(int *array, size_t length, size_t i) {
  size_t major = i;
  size_t left = 2 * i + 1;
  size_t right = 2 * i + 2;

  if (left < length && array[left] > array[major]) {
    major = left;
  }

  if (right < length && array[right] > array[major]) {
    major = right;
  }

  if (i != major) {
    swap(&array[i], &array[major]);
    siftdown(array, length, major);
  }
}

static void heapsort(int *array, size_t length) {
  for (size_t i = length / 2; i > 0; i--) {
    siftdown(array, length, i - 1);
  }

  while (length > 0) {
    length--;
    swap(&array[0], &array[length]);
    siftdown(array, length, 0);
  }
}

static size_t quicksort(int *array, size_t low, size_t high) {
  int pivot = array[high];
  size_t pos = low;

  for (size_t i = 0; i < high; i++) {
    if (array[i] < pivot) {
      swap(&array[i], &array[pos]);
      pos++;
    }
  }

  swap(&array[pos], &array[high]);
  return pos;
}

static void introsort(int *array, size_t length, size_t level) {
  if (length <= INSERTION_SORT_MIN) {
    insertionsort(array, length);
  }
  else if (level == 0) {
    heapsort(array, length);
  }
  else {
    size_t pivot = quicksort(array, 0, length - 1);

    if (pivot > 0) {
      introsort(array, pivot, level - 1);
    }
    introsort(&array[pivot + 1], length - pivot - 1, level - 1);
  }
}

static size_t log2int(size_t val) {
  size_t log = 0;
  while (val > 1) {
    val >>= 1;
    log++;
  }
  return log;
}

void arraylist_sort(ArrayList *list) {
  introsort(list->array, list->elements, log2int(list->elements));
}

void arraylist_reverse(ArrayList *list) {
  if (list->elements <= 1) {
    return;
  }

  for (size_t i = 0; i < list->elements / 2; i++) {
    swap(&list->array[i], &list->array[list->elements - i - 1]);
  }
}

void arraylist_free(ArrayList *list) {
  free(list->array);
  free(list);
}

size_t arraylist_findfirst(ArrayList *list, int val, size_t start, size_t end) {
  if (start > end || end >= list->elements) {
    return 0;
  }

  for (; start <= end; start++) {
    if (val == list->array[start]) {
      return start + 1;
    }
  }

  return 0;
}

size_t arraylist_findlast(ArrayList *list, int val, size_t start, size_t end) {
  if (start > end || end >= list->elements) {
    return 0;
  }

  for (; end >= start; end--) {
    if (val == list->array[end]) {
      return end + 1;
    }
  }

  return 0;
}

size_t arraylist_size(ArrayList *list) {
  return list->elements;
}
